reference/matrix/hybrid_kernels.cpp

```diff
@@ -89,6 +91,71 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_HYBRID_CONVERT_TO_DENSE_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void convert_to_csr(std::shared_ptr<const ReferenceExecutor> exec,
+                    matrix::Csr<ValueType, IndexType> *result,
+                    const matrix::Hybrid<ValueType, IndexType> *source)
+{
+    auto csr_val = result->get_values();
+    auto csr_col_idxs = result->get_col_idxs();
+    auto csr_row_ptrs = result->get_row_ptrs();
+    const auto ell = source->get_ell();
+    const auto max_nnz_per_row = ell->get_num_stored_elements_per_row();
+    const auto coo_val = source->get_const_coo_values();
+    const auto coo_col = source->get_const_coo_col_idxs();
+    const auto coo_row = source->get_const_coo_row_idxs();
+    const auto coo_nnz = source->get_coo_num_stored_elements();
+    csr_row_ptrs[0] = 0;
+    size_type csr_idx = 0;
+    size_type coo_idx = 0;
+    for (IndexType row = 0; row < source->get_size()[0]; row++) {
+        // Ell part
+        for (IndexType col = 0; col < max_nnz_per_row; col++) {
+            const auto val = ell->val_at(row, col);
+            if (val != zero<ValueType>()) {
```

Comment:
Same thing with keeping explicit zeros or not. And further down also.

Reply:
How about we keep explicit zeros for all matrices conversions?
And we also need to implement the function which kicks all zeros out.
Thus, users can decide whether they need to do it.
For reading matrix file now, we seem to delete all zero value, so maybe we also keep them.

Reply:
As far as I know, we should not put the explicit zeros in the ELL format since we might need even more storage that way and I am not sure if we have already have the SpMV improvement in our code that stops as soon as a zero is found.

Actaully, for the Hybrid, I thought it should be fine to ignore all explicit zeros since COO is only used for the parts that don't properly fit into the ELL.

Reply:
For ELL part, it will stops when zero is found. It depends on column index not the value in the cuda kernel.
Hybrid use the nnz_per_row to decide the #col of Ell, so the hybrid matrix is not expected when we skipped the explicit zeros.